### PR TITLE
fix Issue 23145 - Stack allocation of scope new variables defeats @safe

### PR DESCRIPTION
**Pull request by WalterBright**
Timon's idea on how to fix this - disallow constructors for stack allocation for operator new unless the constructor is `scope`.

### CODE REVIEW

**Review by thewilsonator — DISMISSED** [review]
This should target stable unless this is a master only problem.

**Comment by WalterBright** [comment]
This is not a regression, the problem is as old as dip1000.

**Comment by kinke** [comment]
This is primarily a big functional change, definitely worth a big changelog entry. - E.g., I'm pretty sure this prevents all `scope` `Visitor` stack-allocation optimizations in the frontend itself. And requires even more attribute soup to get the previous behavior.

**Comment by WalterBright** [comment]
@kinke you're right. It'll probably have to go behind a preview switch.

**Comment by WalterBright** [comment]
Or perhaps we should just start doing attribute inference for constructors.

**Comment by WalterBright** [comment]
Let's see how far this gets before deciding the best way to deal with older code.

File: src/dmd/nogc.d [hunk 1]
```diff
@@ -140,6 +140,13 @@ public:
         {
             e.error("cannot use `new` in `@nogc` %s `%s`",
                 f.kind(), f.toPrettyChars());
+            if (auto tc = e.type.isTypeClass())
+            {
+                if (e.member && !(e.member.storage_class & STC.scope_))
+                {
+                    e.errorSupplemental("perhaps annotate constructor for `class %s` with `scope`", tc.toChars());
```
Comment (ibuclaw) [inline src/dmd/nogc.d:147]: Deprecation first.

**Review by ibuclaw — DISMISSED** [review]

File: src/dmd/nogc.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/nogc.d]: ok

**Review by thewilsonator — DISMISSED** [review]
This is _not_ a master only issue, please check before making such false assertions. This affects at least 2.099+ (checked on, run.dlang.io, trying all compilers results in a server error).

**Comment by WalterBright** [comment]
I don't understand your point. Putting this in stable will destabilize stable.

**Comment by WalterBright** [comment]
These tests:
```
Failed tests:
test/unit/interfaces/check_implementations_20861.d:100
[Tuple!(string, Kind)("object", private_), Tuple!(string, Kind)("std.stdio", public_), Tuple!(string, Kind)("std.file", private_)]
```
all fail with master on my Ubuntu machine, too. If anyone could find just what code this PR is failing on, I'd appreciate the help.

**Comment by WalterBright** [comment]
More and more it seems the vast bulk of my debugging time is spent just trying to figure out what file where is failing in the test suites. The easy ones to find are the ones in test/runnable, test/compilable, and test/fail_compilation. I dread failures anywhere else, because they are rube goldberg maze.

**Comment by WalterBright** [comment]
circleci fails with:
```
../dmd/generated/linux/debug/64/dmd  -conf= -I../dmd/druntime/import  -w -de -preview=dip1000 -preview=dtorfields -preview=fieldwise -m64 -fPIC -g -debug -lib -ofgenerated/linux/debug/64/libphobos2.a ../dmd/druntime/../generated/linux/debug/64/libdruntime.a std/array.d std/ascii.d std/base64.d std/bigint.d std/bitmanip.d std/checkedint.d std/compiler.d std/complex.d std/concurrency.d std/conv.d std/csv.d std/demangle.d std/encoding.d std/exception.d std/file.d std/functional.d std/getopt.d std/int128.d std/json.d std/mathspecial.d std/meta.d std/mmfile.d std/numeric.d std/outbuffer.d std/package.d std/parallelism.d std/path.d std/process.d std/random.d std/signals.d std/socket.d std/stdint.d std/stdio.d std/string.d std/sumtype.d std/system.d std/traits.d std/typecons.d std/uri.d std/utf.d std/uuid.d std/variant.d std/xml.d std/zip.d std/zlib.d std/algorithm/comparison.d std/algorithm/iteration.d std/algorithm/mutation.d std/algorithm/package.d std/algorithm/searching.d std/algorithm/setops.d std/algorithm/sorting.d std/container/array.d std/container/binaryheap.d std/container/dlist.d std/container/package.d std/container/rbtree.d std/container/slist.d std/container/util.d std/datetime/date.d std/datetime/interval.d std/datetime/package.d std/datetime/stopwatch.d std/datetime/systime.d std/datetime/timezone.d std/digest/crc.d std/digest/hmac.d std/digest/md.d std/digest/murmurhash.d std/digest/package.d std/digest/ripemd.d std/digest/sha.d std/experimental/allocator/common.d std/experimental/allocator/gc_allocator.d std/experimental/allocator/mallocator.d std/experimental/allocator/mmap_allocator.d std/experimental/allocator/package.d std/experimental/allocator/showcase.d std/experimental/allocator/typed.d std/experimental/allocator/building_blocks/affix_allocator.d std/experimental/allocator/building_blocks/aligned_block_list.d std/experimental/allocator/building_blocks/allocator_list.d std/experimental/allocator/building_blocks/ascending_page_allocator.d std/experimental/allocator/building_blocks/bucketizer.d std/experimental/allocator/building_blocks/fallback_allocator.d std/experimental/allocator/building_blocks/free_list.d std/experimental/allocator/building_blocks/free_tree.d std/experimental/allocator/building_blocks/bitmapped_block.d std/experimental/allocator/building_blocks/kernighan_ritchie.d std/experimental/allocator/building_blocks/null_allocator.d std/experimental/allocator/building_blocks/package.d std/experimental/allocator/building_blocks/quantizer.d std/experimental/allocator/building_blocks/region.d std/experimental/allocator/building_blocks/scoped_allocator.d std/experimental/allocator/building_blocks/segregator.d std/experimental/allocator/building_blocks/stats_collector.d std/experimental/logger/core.d std/experimental/logger/filelogger.d std/experimental/logger/nulllogger.d std/experimental/logger/multilogger.d std/experimental/logger/package.d std/format/package.d std/format/read.d std/format/spec.d std/format/write.d std/format/internal/floats.d std/format/internal/read.d std/format/internal/write.d std/math/algebraic.d std/math/constants.d std/math/exponential.d std/math/hardware.d std/math/operations.d std/math/package.d std/math/remainder.d std/math/rounding.d std/math/traits.d std/math/trigonometry.d std/net/curl.d std/net/isemail.d std/uni/package.d std/experimental/checkedint.d std/experimental/typecons.d std/range/interfaces.d std/range/package.d std/range/primitives.d std/regex/package.d std/regex/internal/generator.d std/regex/internal/ir.d std/regex/internal/parser.d std/regex/internal/backtracking.d std/regex/internal/tests.d std/regex/internal/tests2.d std/regex/internal/thompson.d std/regex/internal/kickstart.d std/windows/charset.d std/windows/registry.d std/windows/syserror.d etc/c/curl.d etc/c/odbc/sql.d etc/c/odbc/sqlext.d etc/c/odbc/sqltypes.d etc/c/odbc/sqlucode.d etc/c/sqlite3.d etc/c/zlib.d std/algorithm/internal.d std/digest/digest.d std/internal/cstring.d std/internal/digest/sha_SSSE3.d std/internal/math/biguintcore.d std/internal/math/biguintnoasm.d std/internal/math/biguintx86.d std/internal/math/errorfunction.d std/internal/math/gammafunction.d std/internal/scopebuffer.d std/internal/test/dummyrange.d std/internal/test/range.d std/internal/unicode_comp.d std/internal/unicode_decomp.d std/internal/unicode_grapheme.d std/internal/unicode_norm.d std/internal/unicode_tables.d std/internal/windows/advapi32.d std/typetuple.d generated/linux/debug/64/etc/c/zlib/adler32.o generated/linux/debug/64/etc/c/zlib/compress.o generated/linux/debug/64/etc/c/zlib/crc32.o generated/linux/debug/64/etc/c/zlib/deflate.o generated/linux/debug/64/etc/c/zlib/gzclose.o generated/linux/debug/64/etc/c/zlib/gzlib.o generated/linux/debug/64/etc/c/zlib/gzread.o generated/linux/debug/64/etc/c/zlib/gzwrite.o generated/linux/debug/64/etc/c/zlib/infback.o generated/linux/debug/64/etc/c/zlib/inffast.o generated/linux/debug/64/etc/c/zlib/inflate.o generated/linux/debug/64/etc/c/zlib/inftrees.o generated/linux/debug/64/etc/c/zlib/trees.o generated/linux/debug/64/etc/c/zlib/uncompr.o generated/linux/debug/64/etc/c/zlib/zutil.o
../dmd/generated/linux/debug/64/dmd  -conf= -I../dmd/druntime/import  -w -de -preview=dip1000 -preview=dtorfields -preview=fieldwise -m64 -fPIC -g -debug -shared -defaultlib= -debuglib= -L-lpthread -L-lm -ofgenerated/linux/debug/64/libphobos2.so.0.100.1 -L-soname=libphobos2.so.0.100 ../dmd/druntime/../generated/linux/debug/64/libdruntime.so.a -L-ldl std/array.d std/ascii.d std/base64.d std/bigint.d std/bitmanip.d std/checkedint.d std/compiler.d std/complex.d std/concurrency.d std/conv.d std/csv.d std/demangle.d std/encoding.d std/exception.d std/file.d std/functional.d std/getopt.d std/int128.d std/json.d std/mathspecial.d std/meta.d std/mmfile.d std/numeric.d std/outbuffer.d std/package.d std/parallelism.d std/path.d std/process.d std/random.d std/signals.d std/socket.d std/stdint.d std/stdio.d std/string.d std/sumtype.d std/system.d std/traits.d std/typecons.d std/uri.d std/utf.d std/uuid.d std/variant.d std/xml.d std/zip.d std/zlib.d std/algorithm/comparison.d std/algorithm/iteration.d std/algorithm/mutation.d std/algorithm/package.d std/algorithm/searching.d std/algorithm/setops.d std/algorithm/sorting.d std/container/array.d std/container/binaryheap.d std/container/dlist.d std/container/package.d std/container/rbtree.d std/container/slist.d std/container/util.d std/datetime/date.d std/datetime/interval.d std/datetime/package.d std/datetime/stopwatch.d std/datetime/systime.d std/datetime/timezone.d std/digest/crc.d std/digest/hmac.d std/digest/md.d std/digest/murmurhash.d std/digest/package.d std/digest/ripemd.d std/digest/sha.d std/experimental/allocator/common.d std/experimental/allocator/gc_allocator.d std/experimental/allocator/mallocator.d std/experimental/allocator/mmap_allocator.d std/experimental/allocator/package.d std/experimental/allocator/showcase.d std/experimental/allocator/typed.d std/experimental/allocator/building_blocks/affix_allocator.d std/experimental/allocator/building_blocks/aligned_block_list.d std/experimental/allocator/building_blocks/allocator_list.d std/experimental/allocator/building_blocks/ascending_page_allocator.d std/experimental/allocator/building_blocks/bucketizer.d std/experimental/allocator/building_blocks/fallback_allocator.d std/experimental/allocator/building_blocks/free_list.d std/experimental/allocator/building_blocks/free_tree.d std/experimental/allocator/building_blocks/bitmapped_block.d std/experimental/allocator/building_blocks/kernighan_ritchie.d std/experimental/allocator/building_blocks/null_allocator.d std/experimental/allocator/building_blocks/package.d std/experimental/allocator/building_blocks/quantizer.d std/experimental/allocator/building_blocks/region.d std/experimental/allocator/building_blocks/scoped_allocator.d std/experimental/allocator/building_blocks/segregator.d std/experimental/allocator/building_blocks/stats_collector.d std/experimental/logger/core.d std/experimental/logger/filelogger.d std/experimental/logger/nulllogger.d std/experimental/logger/multilogger.d std/experimental/logger/package.d std/format/package.d std/format/read.d std/format/spec.d std/format/write.d std/format/internal/floats.d std/format/internal/read.d std/format/internal/write.d std/math/algebraic.d std/math/constants.d std/math/exponential.d std/math/hardware.d std/math/operations.d std/math/package.d std/math/remainder.d std/math/rounding.d std/math/traits.d std/math/trigonometry.d std/net/curl.d std/net/isemail.d std/uni/package.d std/experimental/checkedint.d std/experimental/typecons.d std/range/interfaces.d std/range/package.d std/range/primitives.d std/regex/package.d std/regex/internal/generator.d std/regex/internal/ir.d std/regex/internal/parser.d std/regex/internal/backtracking.d std/regex/internal/tests.d std/regex/internal/tests2.d std/regex/internal/thompson.d std/regex/internal/kickstart.d std/windows/charset.d std/windows/registry.d std/windows/syserror.d etc/c/curl.d etc/c/odbc/sql.d etc/c/odbc/sqlext.d etc/c/odbc/sqltypes.d etc/c/odbc/sqlucode.d etc/c/sqlite3.d etc/c/zlib.d std/algorithm/internal.d std/digest/digest.d std/internal/cstring.d std/internal/digest/sha_SSSE3.d std/internal/math/biguintcore.d std/internal/math/biguintnoasm.d std/internal/math/biguintx86.d std/internal/math/errorfunction.d std/internal/math/gammafunction.d std/internal/scopebuffer.d std/internal/test/dummyrange.d std/internal/test/range.d std/internal/unicode_comp.d std/internal/unicode_decomp.d std/internal/unicode_grapheme.d std/internal/unicode_norm.d std/internal/unicode_tables.d std/internal/windows/advapi32.d std/typetuple.d generated/linux/debug/64/etc/c/zlib/adler32.o generated/linux/debug/64/etc/c/zlib/compress.o generated/linux/debug/64/etc/c/zlib/crc32.o generated/linux/debug/64/etc/c/zlib/deflate.o generated/linux/debug/64/etc/c/zlib/gzclose.o generated/linux/debug/64/etc/c/zlib/gzlib.o generated/linux/debug/64/etc/c/zlib/gzread.o generated/linux/debug/64/etc/c/zlib/gzwrite.o generated/linux/debug/64/etc/c/zlib/infback.o generated/linux/debug/64/etc/c/zlib/inffast.o generated/linux/debug/64/etc/c/zlib/inflate.o generated/linux/debug/64/etc/c/zlib/inftrees.o generated/linux/debug/64/etc/c/zlib/trees.o generated/linux/debug/64/etc/c/zlib/uncompr.o generated/linux/debug/64/etc/c/zlib/zutil.o
posix.mak:338: recipe for target 'generated/linux/debug/64/libphobos2.a' failed
```
Nice and useless.

**Comment by kinke** [comment]
> If anyone could find just what code this PR is failing on, I'd appreciate the help.

Not sure if you really just didn't find the corresponding file: https://github.com/dlang/dmd/blob/540d57657f391e84b60ba01af7205622ab0d852e/compiler/test/unit/interfaces/check_implementations_20861.d#L78-L101
[These 'unit' tests use DMD as a library, so it tests an expected error for https://github.com/dlang/dmd/blob/540d57657f391e84b60ba01af7205622ab0d852e/compiler/test/unit/interfaces/check_implementations_20861.d#L83-L93.]

> circleci fails with:
> […]
> Nice and useless.

Only if you ignore the line below that (remember Mathias suggesting to grep for `***` for Makefile errors):
```
make: *** [generated/linux/debug/64/libphobos2.a] Killed
```

After a quick look at CircleCI log and script, it compiles a first compiler, and then uses that one to recompile itself (with coverage). This almost certainly means that the aforementioned regressions for the frontend itself (`scope`-allocated `Visitor` instances) do indeed lead to observable higher memory requirements when compiling Phobos and lead to an out-of-memory error for the 4 GB CircleCI machines (and 2 CPU cores).

File: compiler/src/dmd/nogc.d [hunk 2]
```diff
@@ -138,9 +139,17 @@ public:
             return;                     // separate allocator is called for this, not the GC
         if (f.setGC())
         {
-            e.error("cannot use `new` in `@nogc` %s `%s`",
-                f.kind(), f.toPrettyChars());
-            err = true;
+            if (e.member && !(e.member.storage_class & STC.scope_) && e.type.isTypeClass())
+            {
+                deprecation(e.loc, "cannot use `new` in `@nogc` %s `%s`, perhaps annotate constructor for `class %s` with `scope`",
```
Comment (kinke) [inline compiler/src/dmd/nogc.d:144]: I suggest
a) moving the deprecation out of the `@nogc` violation (so that e.g. the frontend regressions become visible), and
b) retaining the old behavior and instead deprecating the missing `scope` for the constructor, giving people time to adapt their codebases with the extra attribute *without* pessimizing currently working code.

File: compiler/src/dmd/dsymbolsem.d [hunk 3]
```diff
@@ -1037,8 +1037,12 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
...
                         if (auto ne = ex.isNewExp())
                         {
-                            // See if initializer is a NewExp that can be allocated on the stack
-                            if (dsym.type.toBasetype().ty == Tclass)
+                            /* See if initializer is a NewExp that can be allocated on the stack.
+                             * Cannot do it if constructor is not `scope` because the `this` can leak.
+                             * https://issues.dlang.org/show_bug.cgi?id=23145
+                             */
+                            if (dsym.type.toBasetype().ty == Tclass &&
+                                (!ne.member || ne.member.storage_class & STC.scope_))
```
Comment (Geod24) [inline compiler/src/dmd/dsymbolsem.d:1045]: So the compiler will silently heap allocate if I forget `scope` on my constructor ? It should definitely be an error.

File: compiler/src/dmd/nogc.d (hunk 2 above)
Comment (WalterBright) [reply compiler/src/dmd/nogc.d:144]: The trouble with this is it is perfectly legal to "scope-ize" any pointer to anything by giving it a storage class of `scope`. Storing the allocation on the stack is an optimization if the initializer meets additional criteria.

I.e. adding a deprecation for it will break existing code.

**Review by ibuclaw — DISMISSED** [review]
If `scope c = new C` no longer allocates the instance on the stack as a result of this change, then that is an implementation regression and spec violation (and contradicts what we've been telling people to do for decades).

Surely the aim would be not to change codegen, but only to disallow `scope new` in `@safe` code (unless properly annotated)?

**Comment by WalterBright** [comment]
I changed it so it is allowed without comment in unsafe code. In @safe code, a deprecation message is emitted.

**Comment by WalterBright** [comment]
@thewilsonator I still don't understand why you want this in stable.

File: compiler/test/compilable/test23145.d [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    D d;
+    int x=3;
+    this(int i) scope @safe @nogc;
+    this(D d)@safe @nogc;
+}
+
+C foo(D d)@nogc @safe
+{
+    scope e = new C(1);  // ok
+    scope c = new C(d);  // error
```
Comment (ljmf00) [inline compiler/test/compilable/test23145.d:28]: `deprecation` instead of `error`

File: compiler/test/compilable/test23145.d [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+{
+    C c;
+}
+
+class C
+{
+    D d;
+    int x=3;
+    this(int i) scope @safe @nogc;
+    this(D d)@safe @nogc;
```
Comment (ljmf00) [inline compiler/test/compilable/test23145.d:22]: Nit: improve consistency by adding a space before attributes

File: compiler/src/dmd/dsymbolsem.d [hunk 6]
```diff
@@ -1037,9 +1037,24 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
...
                             if (dsym.type.toBasetype().ty == Tclass)
                             {
+                                /* Unsafe to allocate on stack if constructor is not `scope` because the `this` can leak.
+                                 * https://issues.dlang.org/show_bug.cgi?id=23145
+                                 */
+                                if (ne.member && !(ne.member.storage_class & STC.scope_))
+                                {
+                                    if (sc.func.isSafe())
+                                    {
+                                        deprecation(dsym.loc,
```
Comment (ljmf00) [inline compiler/src/dmd/dsymbolsem.d:1051]: Add common deprecation comment to later deprecate after the deprecation period

File: compiler/src/dmd/clone.d [hunk 7]
```diff
@@ -1078,8 +1078,9 @@ private DtorDeclaration buildWindowsCppDtor(AggregateDeclaration ad, DtorDeclara
     Parameter delparam = new Parameter(STC.undefined_, Type.tuns32, Identifier.idPool("del"), new IntegerExp(dtor.loc, 0, Type.tuns32), null);
     Parameters* params = new Parameters;
     params.push(delparam);
-    auto ftype = new TypeFunction(ParameterList(params), Type.tvoidptr, LINK.cpp, dtor.storage_class);
-    auto func = new DtorDeclaration(dtor.loc, dtor.loc, dtor.storage_class, Id.cppdtor);
+    const stc = dtor.storage_class & ~STC.scope_; // because we add the `return this;` later
+    auto ftype = new TypeFunction(ParameterList(params), Type.tvoidptr, LINK.cpp, stc);
+    auto func = new DtorDeclaration(dtor.loc, dtor.loc, stc, Id.cppdtor);
```
Comment (ljmf00) [inline compiler/src/dmd/clone.d:1083]: ThThis is not covered. Perhaps you need to add a test with C++ linkage?

**Review by ljmf00 — DISMISSED** [review]

**Comment by ljmf00** [comment]
Adding Spec PR tag to clarify `scope` attribute applied to constructors.

File: compiler/src/dmd/dsymbolsem.d (hunk 6 above)
Comment (WalterBright) [reply compiler/src/dmd/dsymbolsem.d:1051]: I don't know how long the period should be. Depends on how disruptive this is.

Comment (ljmf00) [reply compiler/src/dmd/dsymbolsem.d:1051]: I think the default is 10 releases, but I indeed believe it's too long. We need to state this in the code before merging to avoid forgetting.

Comment (WalterBright) [reply compiler/src/dmd/dsymbolsem.d:1051]: grepping for  "deprecation" should do the trick.

Comment (ibuclaw) [reply compiler/src/dmd/dsymbolsem.d:1051]: @WalterBright grep for `@@@` and you'll see the comment style used to flag them.

Standard deprecation length is 10 releases, for extreme cases this is 20 releases.

File: compiler/src/dmd/dsymbolsem.d [hunk 8]
```diff
@@ -1037,9 +1037,24 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
...
+                                            "`new` constructor must be annotated with `scope` to initializer `scope` variable `%s`",
```
Comment (benjones) [inline compiler/src/dmd/dsymbolsem.d:1052]: typo: to **initialize** (no r)

File: compiler/src/dmd/nogc.d [hunk 9]
```diff
@@ -13,11 +13,14 @@
...
 module dmd.nogc;
 
+import core.stdc.stdio;
+
 import dmd.aggregate;
 import dmd.apply;
 import dmd.astenums;
 import dmd.declaration;
 import dmd.dscope;
+import dmd.errors;
```
Comment (RazvanN7) [inline compiler/src/dmd/nogc.d:23]: Adding these 2 imports seems to be an unrelated change.

**Review by RazvanN7 — APPROVED** [review]
Mind the nit and please address the other reviewers concerns. @ibuclaw @Geod24 any other objections?

File: compiler/test/compilable/test23145.d [hunk 10]
```diff
@@ -0,0 +1,37 @@
+/* TEST_OUTPUT:
+---
+compilable/test23145.d(117): Deprecation: `new` constructor must be annotated with `scope` to initializer `scope` variable `c`
```
Comment (Geod24) [inline compiler/test/compilable/test23145.d:3]:
```suggestion
compilable/test23145.d(117): Deprecation: `scope` allocation requires that `$FULLY_QUALIFIED_NAME` constructor be annotated with `scope`
```

And an `errorSupplemental` saying "Add `scope` here" with the file/line of the ctor

**Review by Geod24 — COMMENTED** [review]
Suggestion for the message

File: compiler/test/compilable/test23145.d (hunk 10 above)
Comment (RazvanN7) [reply compiler/test/compilable/test23145.d:3]: cc @WalterBright

File: compiler/src/dmd/dsymbolsem.d (hunk 6 above)
Comment (WalterBright) [reply compiler/src/dmd/dsymbolsem.d:1105]: added

File: compiler/src/dmd/nogc.d (hunk 9 above)
Comment (WalterBright) [reply compiler/src/dmd/nogc.d:23]: people keep removing core.stdc.stdio, and I keep adding it back in, because it is for the debugging printf's.

File: compiler/src/dmd/clone.d (hunk 7 above)
Comment (WalterBright) [reply compiler/src/dmd/clone.d:1082]: It is covered, else I never would have thought to add these lines. The coverage tester does not compile Windows specific arms.
